crates/driver/src/boundary/liquidity/uniswap/v3.rs

```diff
@@ -0,0 +1,132 @@
+use {
+    crate::{
+        boundary,
+        domain::{
+            eth,
+            liquidity::{
+                self,
+                uniswap::v3::{Fee, Liquidity, LiquidityNet, Pool, SqrtPrice, Tick},
+            },
+        },
+        infra::{self, blockchain::Ethereum},
+    },
+    anyhow::Result,
+    bigdecimal::ToPrimitive,
+    contracts::{GPv2Settlement, UniswapV3SwapRouter},
+    shared::{
+        current_block::BlockRetrieving,
+        http_solver::model::TokenAmount,
+        interaction::Interaction,
+        sources::uniswap_v3::pool_fetching::UniswapV3PoolFetcher,
+    },
+    solver::{
+        interactions::allowances::Allowances,
+        liquidity::{
+            uniswap_v3::{self, UniswapV3Liquidity, UniswapV3SettlementHandler},
+            ConcentratedLiquidity,
+        },
+        liquidity_collector::LiquidityCollecting,
+    },
+    std::sync::{Arc, Mutex},
+};
+
+pub fn to_domain(id: liquidity::Id, pool: ConcentratedLiquidity) -> liquidity::Liquidity {
+    let handler = pool
+        .settlement_handling
+        .as_any()
+        .downcast_ref::<uniswap_v3::UniswapV3SettlementHandler>()
+        .expect("downcast uniswap settlement handler");
+
+    liquidity::Liquidity {
+        id,
+        gas: eth::Gas(pool.pool.gas_stats.mean_gas),
+        kind: liquidity::Kind::UniswapV3(Pool {
+            router: handler.inner.router.address().into(),
+            address: pool.pool.address.into(),
+            tokens: liquidity::TokenPair::new(
+                pool.pool.tokens[0].id.into(),
+                pool.pool.tokens[1].id.into(),
+            )
+            .unwrap(),
```

Comment:
I wouldn't `unwrap` here but instead return a `None` or `Error`. Specifically, there is no type-level guarantee that `tokens[0].id != tokens[1].id`. The Subgraph could have a bug where the pool token addresses don't get indexed correctly for example and cause us to panic instead of disregard the liquidity.

I think this is different than if you are unwrapping a `boundary::TokenPair` -> `eth::TokenPair` as they both guarantee the same invariants.

Reply:
Made the whole thing fallible: [7803165](https://github.com/cowprotocol/services/pull/1251/commits/78031650880bde643422cd2eea419a388f4e40f8), [ad04836](https://github.com/cowprotocol/services/pull/1251/commits/ad0483642e992c7b938c46dfd3bdf5f1e5709a3e)